### ISSUE TITLE
two questions on zoo attack.

### ISSUE THREAD

**Issue by lihuiknight** [Trusted-AI/adversarial-robustness-toolbox#845]
Dear authors,  I really like this toolbox. it's very helpful.

For the zoo attack on [this class](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/art/attacks/evasion/zoo.py#L50), it's inefficient when there are many examples (vector-form data) to attack.
Why not implement the batch version？

Another question: it seems that coordinate descent is the key limitation for the convergence, is there other method to optimize zeroth order optimization without coordinate descent?

**Comment by beat-buesser** [comment]
Hi @lihuiknight Thank you very much for using ART!

I don't think there is a reason to not implement a batch version of ZooAttack, my guess would be that it was easier to start with implementing a single sample version and because of the option`nb_parallel`, setting the number of coordinate updates running in parallel to speed up the attack, multiplies the batch size. A higher value for `nb_parallel` should be preferable over a large batch size. Are you thinking of implementing a batch version?

I think there are currently no other optimizations implemented in ZooAttack.

**Comment by mathsinn** [comment]
Coordinate-wise descent is indeed query inefficient, we might want to consider using random-vector based version as well. The paper reference: https://openreview.net/forum?id=BJe-DsC5Fm
The more efficient gradient estimator is using the random vectors {u_{i,j}} as defined in (3). i is the index for samples in a batch. If you are attacking images individually, your batch size = 1 and you can ignore the i index. Using (3), gradient estimation takes q+1 queries instead of 2d queries in coordinate case. q is a hyperparameter (the number of random vectors) to balance estimation error and query efficiency. (Thanks to Pin-Yu Chen for the suggestion and guidance.)

**Comment by lihuiknight** [comment]
thanks for sharing this paper. Does the toolbox contain the implementation of this paper?

**Comment by lihuiknight** [comment]
@beat-buesser  thanks for your reply. i notice that `nb_parallel` is used to perform multiple coordinate descent for **ONE** sample with the **vector-form** data. Such approach somehow can accelerate the coordinate descent's convergence. I think when implementation with batch parallel coordinate descent, the code will be more efficient because it can benefit from the tensor tricks in numpy or pytorch.

**Comment by mathsinn** [comment]
> thanks for sharing this paper. Does the toolbox contain the implementation of this paper?

No, it only contains the implementation of the original ZOO algorithm. But we may consider adding this implementation for a future release.
